mlflow/spark.py

```diff
@@ -235,6 +239,9 @@ def log_model(
     try:
         spark_model.save(posixpath.join(model_dir, _SPARK_MODEL_PATH_SUB))
     except Py4JError:
+        if databricks_utils.is_in_cluster() and databricks_utils.is_mlflowdbfs_available():
+            return _log_model_via_mlflowdbfs(run_root_artifact_uri, artifact_path, spark_model)
```

Comment:
Should we return here w/o calling `Model.log()` below?

Reply:
We still need `Model.log()` if `databricks_utils.is_mlflowdbfs_available() == False`.

Reply:
`Model.log()` also makes a call to the MLflow Tracking `LogModel` API automatically; https://github.com/mlflow/mlflow/blob/398345c435421ee2f4ab8d8b3237bc845f25a2c6/mlflow/protos/service.proto#L477-L492.

We should still make sure that this API is called, which means we should fall through to https://github.com/mlflow/mlflow/pull/5759/files#diff-c01988c33f2da6ab63864c712c6258450f6dc8fcedf0566507d0ce3b4ac9ed2bR261-R283 after this portion of the logic executes.

We can accomplish this via:

```
    try:
        spark_model.save(posixpath.join(model_dir, _SPARK_MODEL_PATH_SUB))
    except Py4JError:
        if databricks_utils.is_mlflowdbfs_available():
            _log_model_via_mlflowdbfs(run_root_artifact_uri, artifact_path, spark_model)
        else:
            return Model.log(
                artifact_path=artifact_path,
                flavor=mlflow.spark,
                spark_model=spark_model,
                conda_env=conda_env,
                code_paths=code_paths,
                dfs_tmpdir=dfs_tmpdir,
                sample_input=sample_input,
                registered_model_name=registered_model_name,
                signature=signature,
                input_example=input_example,
                await_registration_for=await_registration_for,
                pip_requirements=pip_requirements,
                extra_pip_requirements=extra_pip_requirements,
            )
```

It would also be great to test that we're saving model metadata when mlflowdbfs is available.

Reply:
Sorry, I don't get the idea. The link above seems to be broken.
The code snippet you suggested does not seem to be calling `Model.log()` if `databricks_utils.is_mlflowdbfs_available()` is true.

Reply:
Right. We shouldn't call `Model.log()` if `databricls_utils.is_mflowdbfs_available()` is `True`. We should instead fall through to the logic at the bottom of the `log_model()` function that replicates the rest of the operations performed by `Model.log()`.

Reply:
Oh, I see! Thanks!